core/mm/core_mmu.c

```diff
@@ -451,28 +504,47 @@ static bool pbuf_is_nsec_ddr(paddr_t pbuf __unused, size_t len __unused)
 #ifdef CFG_SECURE_DATA_PATH
 static bool pbuf_is_sdp_mem(paddr_t pbuf, size_t len)
 {
+	if (sec_sdp.size != 0)
+		return core_is_buffer_inside(pbuf, len,
+					     sec_sdp.paddr,
+					     sec_sdp.size);
+
 	return pbuf_is_special_mem(pbuf, len, phys_sdp_mem_begin,
 				   phys_sdp_mem_end);
 }
 
 struct mobj **core_sdp_mem_create_mobjs(void)
 {
-	const struct core_mmu_phys_mem *mem;
-	struct mobj **mobj_base;
-	struct mobj **mobj;
-	int cnt = phys_sdp_mem_end - phys_sdp_mem_begin;
+	const struct core_mmu_phys_mem *mem = NULL;
+	struct mobj **mobj_base = NULL;
+	struct mobj **mobj = NULL;
+	paddr_t addr = sec_sdp.paddr;
+	paddr_size_t size = sec_sdp.size;
+	int cnt = 1;
+	int i = 0;
 
 	/* SDP mobjs table must end with a NULL entry */
 	mobj_base = calloc(cnt + 1, sizeof(struct mobj *));
```

Comment:
This should be done after `cnt` has been determined.

The logic below isn't very easy to follow. Perhaps it would be better to not mix the two cases in the loop.

Reply:
ok, maybe let's split the logic like that
`struct mobj **core_sdp_mem_create_mobjs(void)
{
	struct mobj **mobj_base = NULL;
	struct mobj **mobj = NULL;
	int i = 0;

	/* SDP mobjs table must end with a NULL entry */
	mobj_base = calloc(cnt + 1, sizeof(struct mobj *));
	if (!mobj_base)
		panic("Out of memory");

	mobj = mobj_base;

	if (!sec_sdp.size) {
		int cnt = phys_sdp_mem_end - phys_sdp_mem_begin;
		const struct core_mmu_phys_mem *mem = phys_sdp_mem_begin;

		for (i = 0; i < cnt; i++) {
			mobj[i] = mobj_phys_alloc(mem[i].addr, mem[i].size,
						  TEE_MATTR_MEM_TYPE_CACHED,
						  CORE_MEM_SDP_MEM);
			if (!mobj[i])
				panic("can't create SDP physical memory object");
		}
	} else {
		mobj[0] = mobj_phys_alloc(sec_sdp.paddr, sec_sdp.size,
					  TEE_MATTR_MEM_TYPE_CACHED,
					  CORE_MEM_SDP_MEM);
		if (!mobj[0])
			panic("can't create SDP physical memory object");
	}

	return mobj_base;
}
`

Reply:
Yes, something like that.

Reply:
done